src/gdisp/gdisp_image_png.c

```diff
@@ -6,1619 +6,165 @@
-		#endif
-
-		for(i = 0; i < d->f.scanbytes; i+=4) {
-			#define pix_color	LUMA2COLOR(d->f.line[i])
-			#define pix_alpha	d->f.line[i+2]
-
-			#if GDISP_NEED_IMAGE_PNG_BACKGROUND
-				if (pix_alpha != 255 && (pinfo->flags & PNG_FLG_BACKGROUND)) {
-					PNG_oColor(&d->o, gdispBlendColor(pix_color, pinfo->bg, pix_alpha));
-					continue;
-				}
-			#endif
-			#if GDISP_NEED_IMAGE_PNG_ALPHACLIFF > 0
-				if (pix_alpha < GDISP_NEED_IMAGE_PNG_ALPHACLIFF) {
-					PNG_oTransparent(&d->o);
-					continue;
-				}
-			#endif
-
-			PNG_oColor(&d->o, pix_color);
+gdispImageError gdispImageOpen_PNG(gdispImage *img) {
+	gdispImagePrivate_PNG *priv = pnglib_file_new(img, 1024);
+	if(priv){
+		img->priv = priv;
+		img->type = GDISP_IMAGE_TYPE_PNG;
 
-			#undef pix_color
-			#undef pix_alpha
+		priv->pr = lib_png_new((lib_reader_read_t) &pnglib_file_read, img);
+		if(priv->pr == NULL){
+			gdispImageClose_PNG(img);
+			return GDISP_IMAGE_ERR_NOMEMORY;
 		}
-	}
-#endif
-#if GDISP_NEED_IMAGE_PNG_RGBALPHA_8
-	static void PNG_OutRGBA8(PNG_decode *d) {
-		unsigned		i;
-		#if GDISP_NEED_IMAGE_PNG_BACKGROUND
-			PNG_info 	*pinfo = d->pinfo;
-		#endif
-
-		for(i = 0; i < d->f.scanbytes; i+=4) {
-			#define pix_color	RGB2COLOR(d->f.line[i+0], d->f.line[i+1], d->f.line[i+2])
-			#define pix_alpha	d->f.line[i+3]
 
-			#if GDISP_NEED_IMAGE_PNG_BACKGROUND
-				if (pix_alpha != 255 && (pinfo->flags & PNG_FLG_BACKGROUND)) {
-					PNG_oColor(&d->o, gdispBlendColor(pix_color, pinfo->bg, pix_alpha));
-					continue;
-				}
-			#endif
-			#if GDISP_NEED_IMAGE_PNG_ALPHACLIFF > 0
-				if (pix_alpha < GDISP_NEED_IMAGE_PNG_ALPHACLIFF) {
-					PNG_oTransparent(&d->o);
-					continue;
-				}
-			#endif
-
-			PNG_oColor(&d->o, pix_color);
-
-			#undef pix_color
-			#undef pix_alpha
+		int header = lib_png_read_header(priv->pr);
+		if(header != 0){
+			gdispImageClose_PNG(img);
+			return GDISP_IMAGE_ERR_BADDATA;
 		}
-	}
-#endif
-#if GDISP_NEED_IMAGE_PNG_RGBALPHA_16
-	static void PNG_OutRGBA16(PNG_decode *d) {
-		unsigned		i;
-		#if GDISP_NEED_IMAGE_PNG_BACKGROUND
-			PNG_info 	*pinfo = d->pinfo;
-		#endif
-
-		for(i = 0; i < d->f.scanbytes; i+=8) {
-			#define pix_color	RGB2COLOR(d->f.line[i+0], d->f.line[i+2], d->f.line[i+4])
-			#define pix_alpha	d->f.line[i+6]
-
-			#if GDISP_NEED_IMAGE_PNG_BACKGROUND
-				if (pix_alpha != 255 && (pinfo->flags & PNG_FLG_BACKGROUND)) {
-					PNG_oColor(&d->o, gdispBlendColor(pix_color, pinfo->bg, pix_alpha));
-					continue;
-				}
-			#endif
-			#if GDISP_NEED_IMAGE_PNG_ALPHACLIFF > 0
-				if (pix_alpha < GDISP_NEED_IMAGE_PNG_ALPHACLIFF) {
-					PNG_oTransparent(&d->o);
-					continue;
-				}
-			#endif
-
-			PNG_oColor(&d->o, pix_color);
-
-			#undef pix_color
-			#undef pix_alpha
+		img->width  = (coord_t) priv->pr->ihdr.width;
+		img->height = (coord_t) priv->pr->ihdr.height;
+		priv->pixbuf_size = img->width * img->height * sizeof(pixel_t);
+		if(priv->pr->ihdr.color_type & 4){
+			img->flags = GDISP_IMAGE_FLG_TRANSPARENT;
```

Comment:
I don't know what this flag does exactly. How does it make the buffer transparent?